packages/@sanity/schema/src/legacy/searchConfig/resolve.ts

```diff
@@ -76,14 +89,16 @@ export function deriveFromPreview(type: {
       weight: PREVIEW_FIELD_WEIGHT_MAP[fieldName],
       path: select[fieldName].split('.'),
     }))
+    .filter((spec) => spec.path.length <= maxDepth)
```

Comment:
nit: We also end up with three iteration on each preview select, we could optimize this by just using a for-loop:

```suggestion
const fields: {weight: number; path: (string | number)[]}[] = []
  for (const fieldName of Object.keys(select)) {
    if (!(fieldName in PREVIEW_FIELD_WEIGHT_MAP)) {
      continue
    }

    const path = select[fieldName].split('.')
    if (maxDepth > -1 && path.length - 1 > maxDepth) {
      continue
    }

    fields.push({
      weight: PREVIEW_FIELD_WEIGHT_MAP[fieldName],
      path,
    })
  }

  return fields
```